xds/internal/xdsclient/xdslbregistry/xdslbregistry_test.go

```diff
@@ -267,51 +294,52 @@ func (s) TestConvertToServiceConfigSuccess(t *testing.T) {
+			wantConfig: jsonMarshal(t, wrrLocalityBalancerConfig(wrrLocalityBalancerConfig(wrrLocalityBalancerConfig(wrrLocalityBalancerConfig(wrrLocalityBalancerConfig(wrrLocalityBalancerConfig(wrrLocalityBalancerConfig(wrrLocalityBalancerConfig(wrrLocalityBalancerConfig(wrrLocalityBalancerConfig(wrrLocalityBalancerConfig(wrrLocalityBalancerConfig(wrrLocalityBalancerConfig(wrrLocalityBalancerConfig(wrrLocalityBalancerConfig(&internalserviceconfig.BalancerConfig{
 				Name: "round_robin",
-			}))))))))))))))),
+			})))))))))))))))),
 		},
 	}
 
 	for _, test := range tests {
 		t.Run(test.name, func(t *testing.T) {
 			if test.rhDisabled {
-				oldRingHashSupport := envconfig.XDSRingHash
+				defer func(old bool) { envconfig.XDSRingHash = old }(envconfig.XDSRingHash)
 				envconfig.XDSRingHash = false
-				defer func() {
-					envconfig.XDSRingHash = oldRingHashSupport
-				}()
+			}
+			if !test.pfDisabled {
+				defer func(old bool) { envconfig.PickFirstLBConfig = old }(envconfig.PickFirstLBConfig)
+				envconfig.PickFirstLBConfig = true
 			}
 			rawJSON, err := xdslbregistry.ConvertToServiceConfig(test.policy, 0)
 			if err != nil {
 				t.Fatalf("ConvertToServiceConfig(%s) failed: %v", pretty.ToJSON(test.policy), err)
 			}
-			bc := &internalserviceconfig.BalancerConfig{}
-			// The converter registry is not guaranteed to emit json that is
-			// valid. It's scope is to simply convert from a proto message to
-			// internal gRPC JSON format. Thus, the tests cause valid JSON to
-			// eventually be emitted from ConvertToServiceConfig(), but this
-			// leaves this test brittle over time in case balancer validations
-			// change over time and add more failure cases. The simplicity of
-			// using this type (to get rid of non determinism in JSON strings)
-			// outweighs this brittleness, and also there are plans on
-			// decoupling the unmarshalling and validation step both present in
-			// this function in the future. In the future if balancer
-			// validations change, any configurations in this test that become
-			// invalid will need to be fixed. (need to make sure emissions above
-			// are valid configuration). Also, once this Unmarshal call is
-			// partitioned into Unmarshal vs. Validation in separate operations,
-			// the brittleness of this test will go away.
-			if err := json.Unmarshal(rawJSON, bc); err != nil {
-				t.Fatalf("failed to unmarshal JSON: %v", err)
+			// got and want must be unmarshalled since JSON strings shouldn't
+			// generally be directly compared.
+			var got []map[string]interface{}
+			if err := json.Unmarshal(rawJSON, &got); err != nil {
+				t.Fatalf("Error unmarshalling rawJSON (%q): %v", rawJSON, err)
+			}
+			var want []map[string]interface{}
+			if err := json.Unmarshal(json.RawMessage(test.wantConfig), &want); err != nil {
```

Comment:
Can you make sure this is WAI by deleting a recursion layer of wrrLocalityBalancerConfig and making sure this fails?